File: backend/src/integrations/ticketing/base_connector.js
/**
 * Base Ticketing Connector — Phase 14.1
 *
 * Defines the vendor-agnostic connector interface for enterprise ticketing systems.
 * Core risk code must remain completely decoupled from vendor-specific payload formatting,
 * status codes, or authentication schemas.
 */

const { TicketRequest, TicketResponse } = require("../../domain/contracts");

class BaseTicketingConnector {
  /**
   * @param {Object} options
   * @param {string} options.name - Unique identifier for connector instance
   * @param {string} options.connectorType - e.g. 'jira', 'servicenow', 'github', 'gitlab', 'webhook'
   * @param {Object} options.config - Vendor-specific configuration (credentials, endpoints)
   * @param {Function} [options.fetchFn] - Optional fetch implementation for testing/mocking
   */
  constructor({ name, connectorType, config = {}, fetchFn = globalThis.fetch } = {}) {
    if (!name) throw new Error("Connector requires a 'name'");
    if (!connectorType) throw new Error("Connector requires a 'connectorType'");
    
    this.name = String(name);
    this.connectorType = String(connectorType).toLowerCase();
    this.config = { ...config };
    this.fetchFn = fetchFn || globalThis.fetch;

    this.validateConfig(this.config);
  }

  /**
   * Validates vendor configuration options. Subclasses should override.
   * @param {Object} config
   */
  validateConfig(config) {
    if (!config) throw new Error(`${this.name} requires configuration object`);
  }

  /**
   * Pure formatter converting a standardized TicketRequest into the vendor's API payload.
   * Subclasses MUST implement this method.
   * @param {TicketRequest} ticketRequest
   * @returns {Object} Vendor-specific payload
   */
  formatPayload(ticketRequest) {
    throw new Error(`formatPayload() must be implemented by ${this.constructor.name}`);
  }

  /**
   * Creates a ticket in the target system.
   * @param {TicketRequest|Object} rawRequest
   * @returns {Promise<TicketResponse>}
   */
  async createTicket(rawRequest) {
    const ticketRequest = rawRequest instanceof TicketRequest
      ? rawRequest
      : new TicketRequest(rawRequest);

    const payload = this.formatPayload(ticketRequest);
    return await this.sendCreateRequest(ticketRequest, payload);
  }

  /**
   * Sends the formatted payload to the vendor API. Subclasses MUST implement.
   * @param {TicketRequest} ticketRequest
   * @param {Object} payload
   * @returns {Promise<TicketResponse>}
   */
  async sendCreateRequest(ticketRequest, payload) {
    throw new Error(`sendCreateRequest() must be implemented by ${this.constructor.name}`);
  }

  /**
   * Tests connectivity and credentials with the vendor API.
   * @returns {Promise<{ ok: boolean, message: string }>}
   */
  async testConnection() {
    return { ok: true, message: `Connection test succeeded for ${this.name}` };
  }

  /**
   * Sanitizes headers/secrets for logging.
   */
  sanitizeHeaders(headers = {}) {
    const sanitized = { ...headers };
    for (const key of Object.keys(sanitized)) {
      if (/(auth|token|secret|password|key)/i.test(key)) {
        sanitized[key] = "[REDACTED]";
      }
    }
    return sanitized;
  }
}

module.exports = {
  BaseTicketingConnector,
};

File: scanners/integrations/base_connector.py
"""
Base Ticketing Connector (Python) — Phase 14.1

Abstract base class defining the vendor-agnostic connector interface.
"""

from abc import ABC, abstractmethod
from typing import Any, Callable, Dict, Optional

from scanners.integrations.ticket_request import TicketRequest, TicketResponse


class BaseTicketingConnector(ABC):
    def __init__(
        self,
        name: str,
        connector_type: str,
        config: Optional[Dict[str, Any]] = None,
        http_client: Optional[Callable[..., Any]] = None,
    ):
        if not name:
            raise ValueError("Connector requires a 'name'")
        if not connector_type:
            raise ValueError("Connector requires a 'connector_type'")

        self.name = str(name)
        self.connector_type = str(connector_type).lower()
        self.config = dict(config or {})
        self.http_client = http_client

        self.validate_config(self.config)

    def validate_config(self, config: Dict[str, Any]) -> None:
        """Validates vendor configuration options. Subclasses should override."""
        if config is None:
            raise ValueError(f"{self.name} requires a configuration dictionary")

    @abstractmethod
    def format_payload(self, ticket_request: TicketRequest) -> Dict[str, Any]:
        """Pure formatter converting TicketRequest into vendor API payload."""
        pass

    @abstractmethod
    def send_create_request(self, ticket_request: TicketRequest, payload: Dict[str, Any]) -> TicketResponse:
        """Sends formatted payload to vendor API."""
        pass

    def create_ticket(self, raw_request: Any) -> TicketResponse:
        """Creates a ticket from TicketRequest or dictionary."""
        if isinstance(raw_request, TicketRequest):
            req = raw_request
        elif isinstance(raw_request, dict):
            req = TicketRequest(**raw_request)
        else:
            raise TypeError("Ticket creation requires TicketRequest or dict")

        payload = self.format_payload(req)
        return self.send_create_request(req, payload)

    def test_connection(self) -> Dict[str, Any]:
        """Tests connectivity and credentials with vendor API."""
        return {"ok": True, "message": f"Connection test succeeded for {self.name}"}
